datasources/firefox/firefox.go

```diff
@@ -0,0 +1,177 @@
+/*
+	Timelinize
+	Copyright (c) 2024 Sergio Rubio
+
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published
+	by the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+// Package firefox implements a data source that imports visited sites
+// from Firefox's places.sqlite database.
+package firefox
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/timelinize/timelinize/timeline"
+	"go.uber.org/zap"
+)
+
+// Data source name and ID.
+const (
+	DataSourceName = "Firefox"
+	DataSourceID   = "firefox"
+)
+
+func init() {
+	err := timeline.RegisterDataSource(timeline.DataSource{
+		Name:            DataSourceID,
+		Title:           DataSourceName,
+		Icon:            "firefox.svg",
+		NewFileImporter: func() timeline.FileImporter { return new(Firefox) },
+	})
+	if err != nil {
+		timeline.Log.Fatal("registering data source", zap.Error(err))
+	}
+}
+
+// Firefox interacts with the places.sqlite database to get visited sites.
+type Firefox struct{}
+
+// Recognize returns whether the input file is recognized.
+func (Firefox) Recognize(_ context.Context, dirEntry timeline.DirEntry, _ timeline.RecognizeParams) (timeline.Recognition, error) {
+	rec := timeline.Recognition{DirThreshold: .9}
+
+	// we can import directories, but let the import planner figure that out; only recognize files
+	if dirEntry.IsDir() {
+		return rec, nil
+	}
```

Comment:
This logic means that you can import an entire directory. To explain:

If you have a folder called "Pictures" (the name is irrelevant though), and it's full of .jpg files, the logic you have here is a good way of telling the import planner: "If 90% of the files in the directory match this data source, we'll take the whole folder." And then later, your FileImport() method would be called on the folder, rather than individual files.

I think this data source only imports a single file (places.sqlite), right?

If so, the logic for both Recognize and FileImport can be much simpler.

Just drop these lines, then in FileImport you can remove the WalkDir().

Let me know if I'm assuming wrong :+1:

Reply:
@mholt no, you're right, although it'd be nice to support importing other kind of files from a firefox profile (such as bookmarks), so importing the whole dir may make sense then, but that'd happen in a future PR.

The changes are still not final though, I was just trying to get it to compile first :wrench:.